txnsync/transactionCache.go

```diff
@@ -0,0 +1,269 @@
+// applies to the timestamp of the most recent transaction in the map.
+const cacheHistoryDuration = 10 * time.Second
+
+// transactionCache is a cache of recently sent transactions ids, allowing to limit the size of the historical kept transactions.
+// transactionCache has FIFO replacement.
+// implementation is a simple cyclic-buffer with a map to accelerate lookups.
+// internally, it's being manages as two tier cache, where the long-term cache is bigger and requires acknowledgements.
+//msgp:ignore transactionCache
+type transactionCache struct {
+	shortTermCache  shortTermTransactionCache
+	longTermCache   longTermTransactionCache
+	ackPendingTxids []ackPendingTxids
+}
+
+//msgp:ignore ackPendingTxids
+type ackPendingTxids struct {
+	txids     []transactions.Txid
+	seq       uint64
+	timestamp time.Duration
+}
+
+//msgp:ignore shortTermTransactionCache
+type shortTermTransactionCache struct {
+	size            int
+	transactionsIDs []transactions.Txid
+	transactionsMap map[transactions.Txid]bool
+	oldest          int
+}
+
+//msgp:ignore longTermTransactionCache
+type longTermTransactionCache struct {
+	current         int
+	transactionsMap []map[transactions.Txid]bool
+	timestamps      []time.Duration
+}
+
+// makeTransactionCache creates the transaction cache
+func makeTransactionCache(shortTermSize, longTermSize, pendingAckTxids int) *transactionCache {
+	txnCache := &transactionCache{
+		shortTermCache: shortTermTransactionCache{
+			size:            shortTermSize,
+			transactionsIDs: make([]transactions.Txid, shortTermSize),
+			transactionsMap: make(map[transactions.Txid]bool, shortTermSize),
+		},
+		ackPendingTxids: make([]ackPendingTxids, 0, pendingAckTxids),
+		longTermCache: longTermTransactionCache{
+			transactionsMap: make([]map[transactions.Txid]bool, (longTermSize+cachedEntriesPerMap-1)/cachedEntriesPerMap),
+			timestamps:      make([]time.Duration, (longTermSize+cachedEntriesPerMap-1)/cachedEntriesPerMap),
+		},
+	}
+	// initialize only the first entry; the rest would be created dynamically.
+	txnCache.longTermCache.transactionsMap[0] = make(map[transactions.Txid]bool, cachedEntriesPerMap)
+	return txnCache
+}
+
+// add adds a single trasaction ID to the short term cache.
+func (lru *transactionCache) add(txid transactions.Txid) {
+	lru.shortTermCache.add(txid)
+}
+
+// addSlice adds a slice to both the short term cache as well as the pending ack transaction ids.
+func (lru *transactionCache) addSlice(txids []transactions.Txid, msgSeq uint64, timestamp time.Duration) {
+	for _, txid := range txids {
+		lru.shortTermCache.add(txid)
+	}
+	// verify that the new msgSeq is bigger than the previous we have.
+	if len(lru.ackPendingTxids) > 0 {
+		if lru.ackPendingTxids[len(lru.ackPendingTxids)-1].seq >= msgSeq {
+			return
+		}
+	}
+
+	if len(lru.ackPendingTxids) == cap(lru.ackPendingTxids) {
+		// roll this array without reallocation.
+		copy(lru.ackPendingTxids, lru.ackPendingTxids[1:])
```

Comment:
why arrays/slices and not lists (that are common for LRU)? In the latter case where would be no full cache copying.

Reply:
I'm not opposing to using a list here, but it would need to be a alloc-free list.

Reply:
as a sidenode, the copy here doesn't really copy any transaction. it copy just the slice object ( without the backing array ).

Reply:
looks like it does - [slicecopy](https://golang.org/src/runtime/slice.go) calls `memmove` that is... well, optimized memmove